### PR TITLE
`Recreate` upgrade strategy using the new Operator

### LINKED ISSUE #13316: `Recreate` upgrade strategy using the new Operator
As a baseline, the Operator should recreate (scale down and up again) the Keycloak deployment during Keycloak upgrade to ensure safe DB migration.

### PR DESCRIPTION
**Pull request by vmuzikar**
Closes #13316

Also sneaked in a small typo fix for probes.

### CODE REVIEW

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeployment.java [hunk 1]
```diff
@@ -633,6 +639,23 @@ public void rollingRestart() {
...
 
+    public void migrateDeployment(StatefulSet previousDeployment, StatefulSet reconciledDeployment) {
+        var previousContainer = previousDeployment.getSpec().getTemplate().getSpec().getContainers().get(0);
+        var reconciledContainer = reconciledDeployment.getSpec().getTemplate().getSpec().getContainers().get(0);
+
+        if (!previousContainer.getImage().equals(reconciledContainer.getImage())
+                && previousDeployment.getStatus().getReplicas() > 1) {
+            // TODO Check if migration is really needed (e.g. based on actual KC version); https://github.com/keycloak/keycloak/issues/10441
+            Log.info("Detected changed Keycloak image, assuming Keycloak upgrade, scaling down the deployment to perform safe DB migration");
+            Log.infof("Orig image: %s; new image: %s");
```
Comment (DGuhr) [inline operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeployment.java:650]:
```suggestion
            Log.infof("original image: %s; new image: %s");
```

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeployment.java [hunk 2]
```diff
@@ -633,6 +639,23 @@ public void rollingRestart() {
...
     }
 
+    public void migrateDeployment(StatefulSet previousDeployment, StatefulSet reconciledDeployment) {
+        var previousContainer = previousDeployment.getSpec().getTemplate().getSpec().getContainers().get(0);
+        var reconciledContainer = reconciledDeployment.getSpec().getTemplate().getSpec().getContainers().get(0);
+
+        if (!previousContainer.getImage().equals(reconciledContainer.getImage())
+                && previousDeployment.getStatus().getReplicas() > 1) {
+            // TODO Check if migration is really needed (e.g. based on actual KC version); https://github.com/keycloak/keycloak/issues/10441
+            Log.info("Detected changed Keycloak image, assuming Keycloak upgrade, scaling down the deployment to perform safe DB migration");
```
Comment (DGuhr) [inline operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeployment.java:649]:
```suggestion
            Log.info("Detected changed Keycloak image, assuming Keycloak upgrade. Scaling down the deployment to one instance to perform a safe database migration");
```

**Review by DGuhr — DISMISSED** [review]
Apart from nitpicks, and assuming the added test is green, LGTM, so approved :) feel free to add the nitpicks or not, it just reads a bit better without all the abbreviations imo.

**Comment by vmuzikar** [comment]
@DGuhr Thanks for the review! Suggestions applied. Would you mind re-approving?

**Review by DGuhr — APPROVED** [review]
@vmuzikar done :)

**Review by DGuhr — CHANGES_REQUESTED** [review]
@vmuzikar tests failing, sadly. can you give it another look pls?

**Comment by vmuzikar** [comment]
@DGuhr Operator CI is green! 🥳

**Review by DGuhr — APPROVED** [review]
approved then :)

**Pull request merged**